lib/govuk_index/popularity_updater.rb

```diff
@@ -0,0 +1,93 @@
+module GovukIndex
+  class PopularityUpdater
+    SCROLL_BATCH_SIZE = 500
+    PROCESSOR_BATCH_SIZE = 100
+    MAX_SIDEKIQ_QUEUE_SIZE = 10
+    MAX_SIDEKIQ_WAIT = 5 * 60
+
+    class SidekiqWaitTimeoutExceeded < StandardError; end
+
+    def self.update(index_name)
+      index_group = SearchConfig.instance.search_server.index_group(index_name)
+      new_index = index_group.create_index
+      index_group.current.with_lock do
+        new(
+          source_index: index_group.current.real_name,
+          destination_index: new_index.real_name
+        ).run
+        index_group.switch_to(new_index)
+      end
+    end
+
+    def initialize(source_index:, destination_index:)
+      @source_index = source_index
+      @destination_index = destination_index
+    end
+
+    def run
+      scroll_enumerator.each_slice(PROCESSOR_BATCH_SIZE) do |documents|
+        wait_for_sidekiq
+        PopularityWorker.perform_async(documents, @destination_index)
+      end
+    end
+
+  private
+
+    def wait_for_sidekiq
+      wait_start = Time.now
+      while Sidekiq::Queue.new.size > MAX_SIDEKIQ_QUEUE_SIZE
+        # alert us that we are waiting.. StatsD?
+        print '.'
+
+        if (Time.now - wait_start) > MAX_SIDEKIQ_WAIT
+          raise SidekiqWaitTimeoutExceeded
+        end
+
+        sleep 1
+      end
+    end
+
+    def scroll_enumerator
+      search_body = { query: { match_all: {} } }
+
+      ScrollEnumerator.new(client: client, index_names: @source_index, search_body: search_body, batch_size: SCROLL_BATCH_SIZE) do |record|
+        {
+          identifier: { _id: record['_id'], _type: record['_type'] },
+          document: record['_source']
+        }
+      end
+    end
+
+    def client
+      GovukIndex::Client.slow_query_instance.client
+    end
+  end
+
+  class PopularityWorker < Indexer::BaseWorker
+    def perform(records, destination_index)
+      actions = ElasticsearchProcessor.new(client: GovukIndex::Client.new(timeout: 30, index_name: destination_index))
+
+      popularities = retrieve_popularities_for(records)
+      records.each do |record|
+        actions.save(
```

Comment:
Instead of `save`, we could add another method to the processor to support the `update` type, which would mean we don't have to send all the data for every document back to elasticsearch.

```
{ "update" : {"_id" : "1", "_type" : "type1", "_index" : "index1"} }
{ "doc" : {"field2" : "value2"} }
```
https://www.elastic.co/guide/en/elasticsearch/reference/1.7/docs-bulk.html

The semantics are documented a bit more here
https://www.elastic.co/guide/en/elasticsearch/reference/1.7/docs-update.html#_updates_with_a_partial_document

> Note, this operation still means full reindex of the document, it just removes some network roundtrips and reduces chances of version conflicts between the get and the index. The _source field need to be enabled for this feature to work.

We'd still have to query ES for the links but we'd be passing much less data around so I'd expect this to help with performance.